torch_ttnn/passes/lowering/add_data_move_pass.py

```diff
@@ -315,252 +287,240 @@ def is_target_a_user_of_curr_node(curr_node, target):
-                    new_args[idx] = new_nodes[-1]
-            dst_node.update_arg(0, tuple(new_args))
+                if old_arg == input_node:
+                    new_args[idx] = aligned_node
+            node.update_arg(0, tuple(new_args))
         else:
-            dst_node.update_arg(dst_idx, new_nodes[-1])
-        return new_nodes[-1]
-    else:
-        return None
+            self._connect_aligned_node(node, aligned_node, input_site, input_site_type)
+
+    def align(self, node, input_node, input_site, input_site_type):
+        assert input_site_type in ["args", "kwargs", "args_tuple", "kwargs_tuple"]
+        align_spec = self._get_align_spec(node, input_node, input_site, input_site_type)
+        if align_spec is None:
+            # No need to align input_node
+            return 0
+        if align_spec in self.aligned_node_dict:
+            aligned_node = self.aligned_node_dict[align_spec]
+        else:
+            with self.graph.inserting_before(node):
+                aligned_node = self._create_aligned_node(align_spec)
+            self.aligned_node_dict[align_spec] = aligned_node
+        if node.target != ttnn.layer_norm:
+            self._connect_aligned_node(node, aligned_node, input_site, input_site_type)
+        else:
+            self._connect_aligned_node_layer_norm(node, input_node, aligned_node, input_site, input_site_type)
+        return 1
 
 
 class AddDataMovePass(PassBase):
     def call(self, gm: torch.fx.GraphModule):
         modified = False
-        device = TtnnDevice()
         i = 0
+        node_input_aligner = NodeInputAligner(gm.graph)
         nodes = list(gm.graph.nodes)
-        # Track argument reuse
-        data_move_in_hash = {}
-        # This might not be needed if workaround is not needed
-        data_move_out_hash = {}
-        for node in nodes:
-            args = node.args[0] if node.op == "output" else node.args
-            kwargs = tuple(_Kwarg(k, v) for k, v in node.kwargs.items() if isinstance(v, torch.fx.node.Node))
-            if isinstance(args, tuple):
-                args += kwargs
 
+        for node in nodes:
+            args = node.args
+            kwargs = node.kwargs
             for idx, arg in enumerate(args):
-                if isinstance(arg, _Kwarg):
-                    try_add_data_move_in_kwargs(arg, node, device)
-                elif (
-                    arg in data_move_in_hash
-                    and not isinstance(node.target, torch._ops.OpOverload)
-                    and node.op != "output"
-                    # TODO: Multiple ops with different layout requirements can have the same input
-                    # There should be a better implementation than skipping certain ops
-                    and node.target not in TTNN_LAYOUT_CHANGE_OPS
-                ):
-                    node.update_arg(idx, data_move_in_hash[arg])
+                if type(arg) not in [tuple, list, torch.fx.immutable_collections.immutable_list]:
```

Comment:
Maybe
```suggestion
                if isinstance(arg, (tuple, list, torch.fx.immutable_collections.immutable_list)):
```

This should also work if arg is a subclass of the classes in the list

Reply:
okok, replaced